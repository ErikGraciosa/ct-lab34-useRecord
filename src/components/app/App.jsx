import React, { useReducer } from 'react';

const initialState = {
  index: 0,
  history: ['#0000FF']
};

function reducer(state, action) {
  switch(action.type) {
    case 'undo':
      return { 
        ...state,
        index: state.index - 1
      };
    case 'redo':
      return { 
        ...state,
        index: state.index + 1
      };
    case 'input': {
      const oldArray = [...state.history];
      const newArray = oldArray.filter((value, index) => index <= state.index);

      return { 
        ...state,
        index: state.index + 1,
        history: [...newArray, action.payload]
      };
    }
    default:
      state;
  }
}

function App() {

  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <>
      <div>Color Changer</div>
      <button onClick={() => dispatch({ type: 'undo' })}>undo</button>
      <button onClick={() => dispatch({ type: 'redo' })}>redo</button>
      <input 
        data-testid="selector"
        name="selector"
        type="color" 
        value={state.history[state.index]} 
        onChange={({ target }) => dispatch({ 
          type: 'input', 
          payload: target.value
        })}
      />
      <div 
        data-testid="output" 
        style={{ 
          backgroundColor: state.history[state.index], 
          width: '10rem', 
          height: '10rem' 
        }}></div>
    </>
  );
}

export default App;
