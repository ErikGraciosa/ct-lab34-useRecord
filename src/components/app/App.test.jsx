import React from 'react';
import { 
  screen,
  render,
  cleanup,
  fireEvent,
  waitFor } from '@testing-library/react';
import App from './App';

describe('App component', () => {
  afterEach(() => cleanup());
  it('renders App', () => {
    const { asFragment } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('verifies that color changes based on input', () => {
    render(<App />);

    const selector = screen.getByTestId('selector');
    const output = screen.getByTestId('output');

    fireEvent.change(selector, {
      target: {
        value: '#FF0000'
      }
    });

    return waitFor(() => {
      expect(output).toHaveStyle({
        backgroundColor: '#FF0000'
      });
    });

  });

  it('verifies that color changes based on undo to previous', () => {
    render(<App />);

    const selector = screen.getByTestId('selector');
    const output = screen.getByTestId('output');
    const undo = screen.getByText('undo');

    fireEvent.change(selector, {
      target: {
        value: '#FF0000'
      }
    });
    fireEvent.change(selector, {
      target: {
        value: '#00FF00'
      }
    });
    fireEvent.change(selector, {
      target: {
        value: '#0000FF'
      }
    });
    fireEvent.click(undo);

    return waitFor(() => {
      expect(output).toHaveStyle({
        backgroundColor: '#00FF00'
      });
    });
  });

  it('verifies color changes based on redo to next index color', () => {
    render(<App />);

    const selector = screen.getByTestId('selector');
    const output = screen.getByTestId('output');
    const undo = screen.getByText('undo');
    const redo = screen.getByText('redo');

    fireEvent.change(selector, {
      target: {
        value: '#FF0000'
      }
    });
    fireEvent.change(selector, {
      target: {
        value: '#00FF00'
      }
    });
    fireEvent.change(selector, {
      target: {
        value: '#0000FF'
      }
    });
    fireEvent.click(undo);
    fireEvent.click(undo);
    fireEvent.click(redo);

    return waitFor(() => {
      expect(output).toHaveStyle({
        backgroundColor: '#00FF00'
      });
    });
  });

});
