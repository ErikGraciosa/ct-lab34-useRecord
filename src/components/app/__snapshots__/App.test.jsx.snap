// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders App 1`] = `
<DocumentFragment>
  <div>
    Color Changer
  </div>
  <button>
    undo
  </button>
  <button>
    redo
  </button>
  <input
    type="color"
    value="#FF0000"
  />
  <div
    style="background-color: rgb(255, 0, 0); width: 10rem; height: 10rem;"
  />
</DocumentFragment>
`;
